src/main/java/org/mockito/internal/junit/ExceptionFactory.java

```diff
@@ -21,21 +21,23 @@ private ExceptionFactory() {
         ExceptionFactoryImpl theFactory = null;
 
         try {
+            Class.forName("org.opentest4j.AssertionFailedError");
             theFactory = new ExceptionFactoryImpl() {
                 @Override
                 public AssertionError create(String message, String wanted, String actual) {
                     return new org.mockito.exceptions.verification.opentest4j.ArgumentsAreDifferent(message, wanted, actual);
                 }
             };
-        } catch (Throwable onlyIfOpenTestIsNotAvailable) {
+        } catch (ClassNotFoundException onlyIfOpenTestIsNotAvailable) {
             try {
+                Class.forName("junit.framework.ComparisonFailure");
                 theFactory = new ExceptionFactoryImpl() {
                     @Override
                     public AssertionError create(String message, String wanted, String actual) {
                         return new org.mockito.exceptions.verification.junit.ArgumentsAreDifferent(message, wanted, actual);
                     }
                 };
-            } catch (Throwable onlyIfJUnitIsNotAvailable) {
+            } catch (ClassNotFoundException onlyIfJUnitIsNotAvailable) {
             }
```

Comment:
something like this may be a little more robust, and also DRYer:

```java
        ExceptionFactoryImpl theFactory = null;

        String[] paths = {
            "org.mockito.exceptions.verification.opentest4j.ArgumentsAreDifferent",
            "org.mockito.exceptions.verification.junit.ArgumentsAreDifferent"
        };
        Constructor<?> constructor = null;
        for(String path : paths) {
            try {
                Class<?> c = Class.forName(path);
                constructor = c.getConstructor(String.class, String.class, String.class);
            } catch (ClassNotFoundException | NoSuchMethodException e) {
                // skip and try next
            }
        }

        if(constructor != null) {
            final Constructor constructor1 = constructor;
            theFactory = (message, wanted, actual) -> {
                try {
                    return (AssertionError) constructor1.newInstance(message, wanted, actual);
                } catch (IllegalAccessException | InstantiationException | InvocationTargetException e) {
                    throw new RuntimeException("failed to instantiate AssertionError", e);
                }
            };
        } else {
            theFactory = (message, wanted, actual) -> {
                return new ArgumentsAreDifferent(message, wanted, actual);
            };
        }
```

Reply:
Hi @tlubz, thanks a bunch for your comments!

Lambdas would certainly be an improvement. I originally wrote it using lambdas and method references, but the main code base was still on 1.6 compliance and so I had to ditch the lambdas (it seems we've since moved on). With lambdas and method references, the code is a lot more compact:

```java
    static {
        ExceptionFactoryImpl theFactory = null;
        try {
            theFactory = org.mockito.exceptions.verification.opentest4j.ArgumentsAreDifferent::new;
        } catch (Throwable t) {
            try {
                theFactory = org.mockito.exceptions.verification.junit.ArgumentsAreDifferent::new;
            } catch (Throwable t2) {
            }
        }
        factory = (theFactory == null) ? ArgumentsAreDifferent::new
                : theFactory;
```

Re: your desire to apply the DRY principle: this is a worthy aim, but because it requires the use of reflection it comes at the expense of compile-time type checking. Both will work, but my personal preference (given the low number of repetitions) was to sacrifice DRY to retain the compile-time type checking. Arguably my compile-time safety "pro" doesn't matter as much because you can catch errors at test time with good unit tests (as we have).

Re: being more robust: this is debatable. Unless you've successfully tried it with `AndroidJUnitRunner`, I suspect that your proposed implementation will still hit issue #1716. The problem is that the actual `org.mockito.exceptions.verification.opentest4j.ArgumentsAreDifferent` and `org.mockito.exceptions.verification.junit.ArgumentsAreDifferent` classes are always found, so they never throw `ClassNotFoundException` when you try and load them. However, when the classloader attempts to *resolve* the class (ie, find all its dependent classes), *this* fails if the appropriate supertype assertion class isn't on the classpath. Failing resolution causes a `NoClassDefFoundError` to be thrown rather than a `ClassNotFoundException`. In a regular JVM environment, this `NoClassDefFoundError` propagates up and is caught by the surrounding `catch (Throwable)`, but it seems that `AndroidJUnitRunner` on the emulator behaves differently - it notices the `NoClassDefFoundError`, logs it, and aborts the test. Hence the patch in this PR tested specifically for the existence of the supertype assertion class (using reflection) before attempted to instantiate the dependent `ArgumentsAreDifferent` class.

So I think for your solution to work, you'd also have to add an explicit test for `org.opentest4j.AssertionError` et al, as this PR did.

Reply:
Great feedback. Thanks for taking the time. mostly what prompted by suggestion is that I was concerned that `Class.forName(...);` may be optimized away by an aggressive optimizer if its result is unused. Though my concerns may be unfounded, since it throws checked exceptions, so it has a potential side-effect. 🤷‍♂